test_data/rules/unused_top_members_in_executable_libraries.dart

```diff
@@ -0,0 +1,56 @@
+// Copyright (c) 2022, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+// test w/ `dart test -N unused_top_members_in_executable_libraries`
+
```

Comment:
Here's a case where we _do_ report an element as unused, even though it is _referenced_: anywhere in the right operand of `is`, _except_ for type aliases:

```dart
class C {}

typedef ListInt = List<int>;

void f(Object p) {
  if (p is C || p is Future<C>) {} // NOT usage
  if (p is ListInt) {} // YES usage
}
```

Reply:
Is this one of the cases you thought would be unrelated? I'm not sure why. In this case, I would want C reported as unused, and I would not want ListInt reported as unused.

Reply:
TBH I don't understand why `C` should be reported here (assuming `f` is reachable from an entrypoint).

The rule doesn't really analyze the execution flow, it only visit the AST children of the `f` node.

Reply:
Perhaps the term _unused_ is a little misleading. It should be understood as _unaccessible from any entry point_.

Reply:
I don't think execution flow is important either; just syntax tree.

In this example, C would be considered _used_, right?

```dart
class C {}

void f() {
  var c = C();
}
```

or even

```dart
class C {}

void f(C c) {}
```

I think those would be "usages" of C. However, in this code:

```dart
class C {}

void f(Object p) {
  if (p is C || p is Future<C>) {}
}
```

I would not count C as being "used." (This is how unused_element works.) Any time `C` is found in the right side of an `is ` expression does not count as usage.

Reply:
> I would not count C as being "used." (This is how unused_element works.) Any time `C` is found in the right side of an `is ` expression does not count as usage.

Thanks for the explanation. 

As mentioned in my in-between comment above my lint don't really mimics the `unused_element` behaviour. It simply visit recursivelly the whole AST (without exception like in the example you provided with right side of `is`) starting from entrypoints and warns if a top public declaration is not reachable.

Reply:
Would `traverseNodesInDFS`'s `excludeCriteria` parameter allow this? I can't tell offhand because I can't find any tests for `traverseNodesInDFS`.

Reply:
I could add this kind of rules but I don't know where to find a list of all the exceptions to mimic the behaviour of `unused_elements`. 
That said I'm not sure the cost of maintaining this logic in 2 places is worth it. If some changes land on `unused_element` we will have to redo the job on linter side.
(Side note: I'm a little surprise that `C` is not considered used because if you remove `class C {}` the code doesn't compile)

Reply:
> Perhaps the term _unused_ is a little misleading. It should be understood as _unaccessible from any entry point_.

If what you're really trying to lint are declarations that are not reachable from `main`, then yes, the terminology is getting in the way. Consider changing the name of the lint to `unreachable_from_main`.

And if that's the case, then we can re-review this PR from that perspective (because it changes the kinds of possible bugs that we should be looking for).

Reply:
> I'm a little surprise that `C` is not considered used because if you remove `class C {}` the code doesn't compile

I have to agree, that does seem a bit odd.

From a tree-shaking perspective, you couldn't remove `C` unless you also removed the then-branch of the `if`. If we think that there can't be any instances of `C` or any subclass of `C`, then it seems like the `if` should be marked as dead code, but `C` shouldn't be reported until the `if` is removed. (But maybe I'm misunderstanding your example.)

Reply:
Sorry for the bad naming :-/ I used this name because of #2395 that has this wording.

> Consider changing the name of the lint to `unreachable_from_main`.

:+1: I will update the PR

Reply:
Renamed. PTAL

Reply:
> I could add this kind of rules but I don't know where to find a list of all the exceptions to mimic the behaviour of unused_elements.

I've just enumerated the exceptions in the docs: https://github.com/dart-lang/sdk/commit/2bedf1e55d593ed88cb63bb9321354ccb6e0bc15

> If we think that there can't be any instances of C or any subclass of C, then it seems like the if should be marked as dead code, ...

Filed https://github.com/dart-lang/sdk/issues/49584

> ... but C shouldn't be reported until the if is removed.

That's fair. We report it today and I don't think we should stop reporting it unless we complete #49584 as it provides value.